File: Class.cpp
#include "Class.h"

using namespace System;
using namespace msclr::interop;

TariffException::TariffException(const std::string& message) :
	std::invalid_argument("TariffException: " + message) {}

PriceStrategy::~PriceStrategy() = default;

double NoDiscountStrategy::calculate_price(double price) const {
	return price;
}

DiscountStrategy::DiscountStrategy(double d) : discount(d) {
	if (discount < 0 || discount > 100) {
		throw TariffException("скидка должна быть вещественным числом в диапазоне от 0 до 100");
	}
}
double DiscountStrategy::calculate_price(double price) const {
	return (1.0 - discount / 100.0) * price;
}

double DiscountStrategy::get_discount() const { return discount; }

Tariff::Tariff(double p, std::string c1, std::string c2) :
	price(p), city1(c1), city2(c2) {
	if (price < 0) {
		throw TariffException("цена не может быть отрицательной");
	}
}
Tariff::~Tariff() = default;
double Tariff::get_base_price() const { return price; }
std::string Tariff::get_city1() const { return city1; }
std::string Tariff::get_city2() const { return city2; }

UsualTariff::UsualTariff(double p, std::string c1, std::string c2) :
	Tariff(p, c1, c2) {
}

DiscountTariff::DiscountTariff(double p, double discount, std::string c1, std::string c2) :
	Tariff(p, c1, c2), DiscountStrategy{ discount } {
}

void ATE::add_usual_tariff(double price, const std::string& city1, const std::string& city2) {
	usual_tariffs.push_back(new UsualTariff(price, city1, city2));
}
void ATE::add_discount_tariff(double price, double discount, const std::string& city1, const std::string& city2) {
	discount_tariffs.push_back(new DiscountTariff(price, discount, city1, city2));
}
ATE& ATE::operator += (const UsualTariff& tariff) {
	total += tariff.calculate_price(tariff.get_base_price());
	return *this;
}
ATE& ATE::operator += (const DiscountTariff& tariff) {
	total += tariff.calculate_price(tariff.get_base_price());
	return *this;
}
double ATE::calculate_average_price() {
	if (usual_tariffs.empty() && discount_tariffs.empty()) {
		throw TariffException("списки тарифов пусты");
	}
	for (int i = 0; i < usual_tariffs.size(); i++) {
		*this += *usual_tariffs[i];
	}
	for (int i = 0; i < discount_tariffs.size(); i++) {
		*this += *discount_tariffs[i];
	}
	return total / (double)(size(usual_tariffs) + size(discount_tariffs));
}

std::vector<std::string> ATE::Split(std::string str, char sep) {
	std::stringstream ss(str);
	std::vector<std::string> tokens;
	std::string token;
	while (std::getline(ss, token, sep)) {
		tokens.push_back(token);
	}
	return tokens;
}

void ATE::SetLists(System::Collections::Generic::List<System::String^>^ tariff_list) {
	for (auto* tariff : usual_tariffs) {
		delete tariff;
	}
	usual_tariffs.clear();
	for (auto* tariff : discount_tariffs) {
		delete tariff;
	}
	discount_tariffs.clear();
	for (int i = 0; i < tariff_list->Count; i++) {
		std::string line = marshal_as<std::string>(tariff_list[i]);
		std::vector<std::string> parts = Split(line, ';');
		if (parts.size() == 4 && parts[0] == "u") {
			std::string city1 = parts[1];
			std::string city2 = parts[2];
			double price = stod(parts[3]);
			usual_tariffs.push_back(new UsualTariff(price, city1, city2));
		}
		else if (parts.size() == 5 && parts[0] == "d") {
			std::string city1 = parts[1];
			std::string city2 = parts[2];
			double price = stod(parts[3]);
			double discount = stod(parts[4]);
			discount_tariffs.push_back(new DiscountTariff(price, discount, city1, city2));
		}
		else {
			throw TariffException("Неверный формат строки в списке тарифов.");
		}
	}
}

int ATE::CompareTariffsByPrice(System::String^ x, System::String^ y) {
	try {
		array<System::String^>^ parts_x = x->Split(gcnew array<System::Char>{L';'});
		array<System::String^>^ parts_y = y->Split(gcnew array<System::Char>{L';'});
		if (parts_x->Length < 4 || parts_x->Length > 5 || parts_y->Length < 4 || parts_y->Length > 5)
			throw TariffException("Неверный формат строки в списке тарифов.");
		if (Convert::ToDouble(parts_x[3]) < Convert::ToDouble(parts_y[3])) {
			return -1;
		}
		else if (Convert::ToDouble(parts_x[3]) > Convert::ToDouble(parts_y[3])) {
			return 1;
		}
		else {
			return 0;
		}
	}
	catch (...) {
		throw TariffException("Ошибка при сортировке тарифов по цене.");
	}

}

int ATE::CompareTariffsByCity1(System::String^ x, System::String^ y) {
	try {
		array<System::String^>^ parts_x = x->Split(gcnew array<System::Char>{L';'});
		array<System::String^>^ parts_y = y->Split(gcnew array<System::Char>{L';'});
		if (parts_x->Length < 4 || parts_x->Length > 5 || parts_y->Length < 4 || parts_y->Length > 5)
			throw TariffException("Неверный формат строки в списке тарифов.");
		return System::String::Compare(parts_x[1], parts_y[1]);
	} 
	catch (...) {
		throw TariffException("Ошибка при сортировке тарифов по цене.");
	}
}

int ATE::CompareTariffsByCity2(System::String^ x, System::String^ y) {
	try {
		array<System::String^>^ parts_x = x->Split(gcnew array<System::Char>{L';'});
		array<System::String^>^ parts_y = y->Split(gcnew array<System::Char>{L';'});
		if (parts_x->Length < 4 || parts_x->Length > 5 || parts_y->Length < 4 || parts_y->Length > 5)
			throw TariffException("Неверный формат строки в списке тарифов.");
		return System::String::Compare(parts_x[2], parts_y[2]);
	}
	catch (...) {
		throw TariffException("Ошибка при сортировке тарифов по цене.");
	}
}

System::Collections::Generic::List<System::String^>^ ATE::SortTariffsByPrice(System::Collections::Generic::List<System::String^>^ tariff_list) {
	System::Collections::Generic::List<System::String^>^ sorted_list = gcnew System::Collections::Generic::List<System::String^>(tariff_list);
	sorted_list->Sort(gcnew System::Comparison<System::String^>(ATE::CompareTariffsByPrice));
	return sorted_list;
}

System::Collections::Generic::List<System::String^>^ ATE::SortTariffsByCity1(System::Collections::Generic::List<System::String^>^ tariff_list) {
	System::Collections::Generic::List<System::String^>^ sorted_list = gcnew System::Collections::Generic::List<System::String^>(tariff_list);
	sorted_list->Sort(gcnew System::Comparison<System::String^>(ATE::CompareTariffsByCity1));
	return sorted_list;
}

System::Collections::Generic::List<System::String^>^ ATE::SortTariffsByCity2(System::Collections::Generic::List<System::String^>^ tariff_list) {
	System::Collections::Generic::List<System::String^>^ sorted_list = gcnew System::Collections::Generic::List<System::String^>(tariff_list);
	sorted_list->Sort(gcnew System::Comparison<System::String^>(ATE::CompareTariffsByCity2));
	return sorted_list;
}

File: Class.cs
using Microsoft.Data.Sqlite;
using System.Globalization;

public class TariffException : ArgumentException
{
    public TariffException(string message) : base(message) { }
}

public interface IPriceStrategy
{
    double CalculatePrice(double price);
}

public class NoDiscountStrategy : IPriceStrategy
{
    public double CalculatePrice(double price) => price;
}

public class DiscountStrategy : IPriceStrategy
{
    public double Discount { get; }

    public DiscountStrategy(double discount)
    {
        if (discount < 0 || discount > 100)
            throw new TariffException("скидка должна быть вещественным числом в диапазоне от 0 до 100");
        Discount = discount;
    }
    public double CalculatePrice(double price) => price - price * Discount / 100.0;
}

public abstract class Tariff
{ 
    public double Price { get; }
    public string City1 { get; }
    public string City2 { get; }

    protected Tariff(double price, string c1, string c2)
    {
        if (price < 0)
            throw new TariffException("цена не может быть отрицательной");
        Price = price;
        City1 = c1;
        City2 = c2;
    }

    public abstract double GetPrice();
}

public class UsualTariff : Tariff
{
    public readonly NoDiscountStrategy Strategy = new NoDiscountStrategy();
    public UsualTariff(double price, string c1, string c2) : base(price, c1, c2) { }
    override public double GetPrice() => Strategy.CalculatePrice(Price);
}

public class DiscountTariff : Tariff
{
    public double Discount;
    public readonly DiscountStrategy Strategy;
    public DiscountTariff(double price, double discount, string c1, string c2) : base(price, c1, c2)
    {
        Strategy = new DiscountStrategy(discount);
        Discount = discount;
    }
    override public double GetPrice() => Strategy.CalculatePrice(Price);
}

public class ATE
{
    private readonly List<UsualTariff> UsualTariffs = new List<UsualTariff>();
    private readonly List<DiscountTariff> DiscountTariffs = new List<DiscountTariff>();
    public void AddUsualTariff(double p, string c1, string c2)
    {
        UsualTariffs.Add(new UsualTariff(p, c1, c2));
    }
    public void AddDiscountTariff(double p, double d, string c1, string c2)
    {
        DiscountTariffs.Add(new DiscountTariff(p, d, c1, c2));
    }
    public double CalculateAveragePrice()
    {
        if (UsualTariffs.Count == 0 && DiscountTariffs.Count == 0)
            throw new TariffException("списки тарифов пусты");
        return UsualTariffs.Sum(l => l.GetPrice()) + DiscountTariffs.Sum(l => l.GetPrice()) / (double)(UsualTariffs.Count + DiscountTariffs.Count);
    }
    public void SetLists(List<string> tariff_list)
    {
        UsualTariffs.Clear();
        DiscountTariffs.Clear();
        foreach (string line in tariff_list)
        {
            string[] parts = line.Split(";");
            if (parts.Length == 4 && parts[0] == "u")
            {
                string c1 = parts[1];
                string c2 = parts[2];
                double p = double.Parse(parts[3], CultureInfo.InvariantCulture);
                UsualTariffs.Add(new UsualTariff(p, c1, c2));
            }
            else if (parts.Length == 5 && parts[0] == "d")
            {
                string c1 = parts[1];
                string c2 = parts[2];
                double p = double.Parse(parts[3], CultureInfo.InvariantCulture);
                double d = double.Parse(parts[4], CultureInfo.InvariantCulture);
                DiscountTariffs.Add(new DiscountTariff(p, d, c1, c2));
            }
            else
            {
                throw new TariffException("неверный формат строки в списке тарифов");
            }
        }
    }
    public List<string> GetLists()
    {
        var tariff_list = new List<string>();
        foreach (var t in UsualTariffs)
            tariff_list.Add($"u;{t.City1};{t.City2};{t.Price.ToString("F2", CultureInfo.InvariantCulture)}");
        foreach (var t in DiscountTariffs)
            tariff_list.Add($"d;{t.City1};{t.City2};{t.Price.ToString("F2", CultureInfo.InvariantCulture)};{t.Discount.ToString("F2", CultureInfo.InvariantCulture)}");
        return tariff_list;
    }
    public List<string> SortByPrice(List<string> tariff_list)
    {
        return tariff_list.OrderBy(line => double.Parse(line.Split(';')[3], CultureInfo.InvariantCulture)).ToList();
    }
    public List<string> SortByCity1(List<string> tariff_list)
    {
        return tariff_list.OrderBy(line => line.Split(';')[1]).ToList();
    }
    public List<string> SortByCity2(List<string> tariff_list)
    {
        return tariff_list.OrderBy(line => line.Split(';')[2]).ToList();
    }

    public string dbPath;
    public bool dbMode { get; set; }

    public bool InitializeDatabase(string db_path)
    {
        try
        {
            dbPath = db_path;
            using var connection = new SqliteConnection($"Data Source={dbPath}");
            connection.Open();
            using var command = connection.CreateCommand();
            command.CommandText = @"
            CREATE TABLE IF NOT EXISTS Tariffs (
                Id INTEGER PRIMARY KEY AUTOINCREMENT,
                City1 TEXT NOT NULL,
                City2 TEXT NOT NULL,
                Price REAL NOT NULL,
                Discount REAL,
                Type TEXT CHECK(Type IN ('Usual', 'Discount')) NOT NULL
            )";
            command.ExecuteNonQuery();
            dbMode = true;
            return true;
        }
        catch
        {
            dbMode = false;
            return false;
        }
    }

    public void SaveToDatabase()
    {
        if (!dbMode) return;
        using var connection = new SqliteConnection($"Data Source={dbPath}");
        connection.Open();
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "DELETE FROM Tariffs";
            command.ExecuteNonQuery();
        }
        foreach (var tariff in UsualTariffs)
        {
            using var command = connection.CreateCommand();
            command.CommandText = @"
                INSERT INTO Tariffs (City1, City2, Price, Discount, Type)
                VALUES ($city1, $city2, $price, NULL, 'Usual')";
            command.Parameters.AddWithValue("$city1", tariff.City1);
            command.Parameters.AddWithValue("$city2", tariff.City2);
            command.Parameters.AddWithValue("$price", tariff.Price);
            command.ExecuteNonQuery();
        }
        foreach (var tariff in DiscountTariffs)
        {
            using var command = connection.CreateCommand();
            command.CommandText = @"
                INSERT INTO Tariffs (City1, City2, Price, Discount, Type)
                VALUES ($city1, $city2, $price, $discount, 'Usual')";
            command.Parameters.AddWithValue("$city1", tariff.City1);
            command.Parameters.AddWithValue("$city2", tariff.City2);
            command.Parameters.AddWithValue("$price", tariff.Price);
            command.Parameters.AddWithValue("$discount", tariff.Discount);
            command.ExecuteNonQuery();
        }
    }
    public void LoadFromDatabase()
    {
        if (!dbMode) return;
        UsualTariffs.Clear();
        DiscountTariffs.Clear();
        using var connection = new SqliteConnection($"Data Source={dbPath}");
        connection.Open();
        using var command = connection.CreateCommand();
        command.CommandText = "SELECT City1, City2, Price, Discount, Type FROM Tariffs";
        using var reader = command.ExecuteReader();
        while (reader.Read())
        {
            string city1 = reader.GetString(0);
            string city2 = reader.GetString(1);
            double price = reader.GetDouble(2);
            string type = reader.GetString(4);
            if (type == "Usual")
            {
                UsualTariffs.Add(new UsualTariff(price, city1, city2));
            } else if (type == "Discount")
            {
                double discount = reader.IsDBNull(3) ? 0.0 : reader.GetDouble(3);
                DiscountTariffs.Add(new DiscountTariff(price, discount, city1, city2));
            }
        }
    }
}